File: fastapi_app/lib/doi_utils.py
"""
DOI metadata utilities for fetching and parsing bibliographic information.

Provides:
- DOI validation and normalization
- Filesystem-safe encoding/decoding for doc_ids
- Metadata fetching from CrossRef and DataCite APIs
"""

import re
import requests
from typing import Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)

# DOI validation regex (from CrossRef specification)
# Matches: 10.{4-9 digits}/{suffix with allowed characters}
# Allowed suffix characters: A-Z 0-9 -._;()/
DOI_REGEX = r"^10\.\d{4,9}/[-._;()/:A-Z0-9]+$"


def validate_doi(doi: str) -> bool:
    """
    Check if a DOI string is valid according to CrossRef specifications.

    Args:
        doi: The DOI string to validate

    Returns:
        True if DOI matches the valid pattern, False otherwise

    Examples:
        >>> validate_doi("10.5771/2699-1284-2024-3-149")
        True
        >>> validate_doi("not-a-doi")
        False
        >>> validate_doi("10.1234/valid-doi_123")
        True
    """
    if not doi:
        return False
    return bool(re.match(DOI_REGEX, doi, flags=re.IGNORECASE))


def fetch_doi_metadata(doi: str, timeout: int = 10) -> Dict[str, Any]:
    """
    Fetch metadata for DOI from CrossRef or DataCite APIs.

    Tries CrossRef first (most common), falls back to DataCite.

    Args:
        doi: The DOI string to fetch metadata for
        timeout: Request timeout in seconds (default: 10)

    Returns:
        Dictionary with metadata fields:
            - title: Article/document title
            - authors: List of dicts with 'given' and 'family' name
            - date: Publication year
            - publisher: Publisher name
            - journal: Journal/container title
            - volume: Volume number
            - issue: Issue number
            - pages: Page range

    Raises:
        ValueError: If DOI format is invalid
        requests.exceptions.HTTPError: If both APIs fail
        requests.exceptions.Timeout: If request times out

    Examples:
        >>> metadata = fetch_doi_metadata("10.5771/2699-1284-2024-3-149")
        >>> metadata['title']
        "Legal status of Derived Text Formats..."
    """
    if not validate_doi(doi):
        raise ValueError(f"{doi} is not a valid DOI string")

    # Try CrossRef first (more common for academic papers)
    try:
        logger.debug(f"Fetching metadata from CrossRef for DOI: {doi}")
        return parse_crossref_metadata(doi, timeout)
    except requests.exceptions.HTTPError as e:
        logger.warning(f"CrossRef API failed for {doi}: {e}")
        # Fall back to DataCite
        try:
            logger.debug(f"Falling back to DataCite for DOI: {doi}")
            return parse_datacite_metadata(doi, timeout)
        except requests.exceptions.HTTPError as e2:
            logger.error(f"Both CrossRef and DataCite APIs failed for {doi}")
            raise


def parse_crossref_metadata(doi: str, timeout: int = 10) -> Dict[str, Any]:
    """
    Parse metadata from CrossRef API.

    Args:
        doi: The DOI to look up
        timeout: Request timeout in seconds

    Returns:
        Dictionary with parsed metadata

    Raises:
        requests.exceptions.HTTPError: If API request fails
    """
    url = f"https://api.crossref.org/works/{doi}"
    response = requests.get(url, timeout=timeout)
    response.raise_for_status()
    data = response.json()

    message = data.get("message", {})

    # Extract title
    title = message.get("title", [None])[0]

    # Extract authors
    authors_data = message.get("author", [])
    authors = [
        {"given": author.get("given"), "family": author.get("family")}
        for author in authors_data
    ]

    # Extract publication date
    issued_data = message.get("issued", {})
    date_parts = issued_data.get("date-parts", [[]])
    date = date_parts[0][0] if date_parts and date_parts[0] else None

    return {
        "title": title,
        "authors": authors,
        "date": date,
        "publisher": message.get("publisher"),
        "journal": message.get("container-title", [None])[0],
        "volume": message.get("volume"),
        "issue": message.get("issue"),
        "pages": message.get("page"),
    }


def parse_datacite_metadata(doi: str, timeout: int = 10) -> Dict[str, Any]:
    """
    Parse metadata from DataCite API.

    Args:
        doi: The DOI to look up
        timeout: Request timeout in seconds

    Returns:
        Dictionary with parsed metadata

    Raises:
        requests.exceptions.HTTPError: If API request fails
    """
    url = f"https://api.datacite.org/dois/{doi}"
    response = requests.get(url, timeout=timeout)
    response.raise_for_status()
    data = response.json()

    attributes = data.get("data", {}).get("attributes", {})

    # Extract title
    title = None
    if attributes.get("titles"):
        title = attributes["titles"][0].get("title")

    # Extract authors (creators in DataCite)
    authors_data = attributes.get("creators", [])
    authors = [
        {"given": author.get("givenName"), "family": author.get("familyName")}
        for author in authors_data
    ]

    # Extract journal/container
    journal = None
    if attributes.get("container"):
        journal = attributes["container"].get("title")

    return {
        "title": title,
        "authors": authors,
        "date": attributes.get("publicationYear"),
        "publisher": attributes.get("publisher"),
        "journal": journal,
        "volume": attributes.get("volume", ""),
        "issue": attributes.get("issue", ""),
        "pages": attributes.get("page"),
    }


def normalize_doi(doi: str) -> str:
    """
    Normalize a DOI string.

    - Removes leading/trailing whitespace
    - Removes common prefixes (doi:, http://doi.org/, https://doi.org/)
    - Converts to lowercase (DOIs are case-insensitive)

    Args:
        doi: DOI string to normalize

    Returns:
        Normalized DOI string

    Examples:
        >>> normalize_doi("  10.5771/2699-1284-2024-3-149  ")
        "10.5771/2699-1284-2024-3-149"
        >>> normalize_doi("doi:10.5771/2699-1284-2024-3-149")
        "10.5771/2699-1284-2024-3-149"
        >>> normalize_doi("https://doi.org/10.5771/2699-1284-2024-3-149")
        "10.5771/2699-1284-2024-3-149"
    """
    if not doi:
        return doi

    # Strip whitespace
    doi = doi.strip()

    # Remove common DOI prefixes
    prefixes = [
        "doi:",
        "DOI:",
        "http://doi.org/",
        "https://doi.org/",
        "http://dx.doi.org/",
        "https://dx.doi.org/",
    ]

    for prefix in prefixes:
        if doi.startswith(prefix):
            doi = doi[len(prefix):]
            break

    # DOIs are case-insensitive, but conventionally lowercase
    # However, keep original case for compatibility
    return doi


# Filesystem encoding utilities
# ==============================

def is_filename_encoded(filename: str) -> bool:
    """
    Check if a filename is already encoded.

    Detects encoding markers:
    - Double underscore (__) where original had forward slash
    - Dollar-sign patterns ($XX$) for special characters

    Args:
        filename: Filename to check

    Returns:
        True if filename appears to be already encoded, False otherwise

    Examples:
        >>> is_filename_encoded("10.1111__1467-6478.00040")
        True
        >>> is_filename_encoded("10.1234__test$3A$file")
        True
        >>> is_filename_encoded("10.1111/1467-6478.00040")
        False
        >>> is_filename_encoded("simple-filename.txt")
        False
    """
    # Check for $XX$ encoding pattern
    if re.search(r'\$[0-9A-F]{2}\$', filename):
        return True

    # Check for __ (encoded slash) combined with DOI-like pattern
    # If it has __ and looks like a DOI prefix, it's likely encoded
    if '__' in filename and re.match(r'^10\.\d+__', filename):
        return True

    return False


def encode_filename(doc_id: str) -> str:
    """
    Encode a document ID (e.g., DOI) to a filesystem-safe filename.

    Encoding rules:
    - Forward slashes (/) → double underscore (__)
    - Other filesystem-incompatible characters → dollar-sign encoding ($XX$)
      where XX is the hexadecimal representation of the character code
    - Dollar sign is used instead of percent for better cross-platform compatibility

    Args:
        doc_id: Document identifier to encode (e.g., DOI, file reference)

    Returns:
        Filesystem-safe encoded string

    Raises:
        ValueError: If doc_id is empty

    Examples:
        >>> encode_filename("10.1111/1467-6478.00040")
        "10.1111__1467-6478.00040"
        >>> encode_filename("10.1234/test:file")
        "10.1234__test$3A$file"
        >>> encode_filename("doc<name>")
        "doc$3C$name$3E$"
    """
    if not doc_id:
        raise ValueError("doc_id cannot be empty")

    # First handle forward slashes specially (common in DOIs)
    result = doc_id.replace("/", "__")

    # Characters that are unsafe on various filesystems
    # Windows: < > : " | ? * and control chars
    # Unix/Mac: / (already handled) and control chars
    # We'll encode anything that might be problematic
    unsafe_chars = set('<>:"|?*\\')

    # Also encode dollar sign itself to avoid ambiguity
    unsafe_chars.add('$')

    # Build encoded string
    encoded = []
    for char in result:
        if char in unsafe_chars or ord(char) < 32:
            # Encode as $XX$ where XX is hex
            encoded.append(f"${ord(char):02X}$")
        else:
            encoded.append(char)

    return ''.join(encoded)


def decode_filename(filename: str) -> str:
    """
    Decode a filesystem-safe filename back to the original document ID.

    Reverses the encoding applied by encode_filename().
    Handles both new dollar-sign encoding and legacy formats.

    Args:
        filename: Encoded filename to decode

    Returns:
        Original document ID string

    Raises:
        ValueError: If filename is empty or contains invalid encoding

    Examples:
        >>> decode_filename("10.1111__1467-6478.00040")
        "10.1111/1467-6478.00040"
        >>> decode_filename("10.1234__test$3A$file")
        "10.1234/test:file"
        >>> decode_filename("doc$3C$name$3E$")
        "doc<name>"
    """
    if not filename:
        raise ValueError("filename cannot be empty")

    # First, restore forward slashes from double underscores
    result = filename.replace("__", "/")

    # Decode $XX$ patterns
    decoded = []
    i = 0
    while i < len(result):
        if result[i] == '$':
            # Look for pattern $XX$
            if i + 3 < len(result) and result[i+3] == '$':
                hex_code = result[i+1:i+3]
                try:
                    char_code = int(hex_code, 16)
                    decoded.append(chr(char_code))
                    i += 4
                    continue
                except ValueError:
                    raise ValueError(f"Invalid hex encoding: ${hex_code}$")
            else:
                raise ValueError(f"Invalid encoding at position {i}: incomplete $XX$ pattern")
        else:
            decoded.append(result[i])
            i += 1

    return ''.join(decoded)

File: app/src/modules/doi-utils.js
/**
 * DOI utility functions mirroring fastapi_app/lib/utils/doi_utils.py.
 *
 * Provides:
 * - DOI validation and normalization
 * - Filesystem-safe encoding/decoding for doc_ids
 *
 * Keep in sync with doi_utils.py.
 */

// DOI validation regex (from CrossRef specification)
// Matches: 10.{4-9 digits}/{suffix with allowed characters}
export const DOI_REGEX = /^10\.\d{4,9}\/[-._;()\/:A-Z0-9]+$/i;

const DOI_EXTRACT_PATTERN = /10\.\d{4,9}\/[-._;()\/:A-Z0-9]+/i;

const _LEGACY_ENCODING_RE = /\$([0-9A-Fa-f]{2})\$/g;


/**
 * Convert legacy $XX$ percent-style encoding to the current _xXX_ format.
 * Mirrors: normalize_legacy_encoding()
 * @param {string} s
 * @returns {string}
 */
export function normalizeLegacyEncoding(s) {
  return s.replace(_LEGACY_ENCODING_RE, (_, hex) => `_x${hex.toUpperCase()}_`)
}


/**
 * Check if a filename is already encoded.
 * Mirrors: is_filename_encoded()
 * @param {string} filename
 * @returns {boolean}
 */
export function isFilenameEncoded(filename) {
  if (!filename) return false
  if (/_x[0-9A-F]{2}_/i.test(filename)) return true
  if (/\$[0-9A-F]{2}\$/i.test(filename)) return true
  if (filename.includes('__') && /^10\.\d+__/.test(filename)) return true
  return false
}


/**
 * Encode a document ID (e.g. DOI) to a filesystem-safe filename.
 *
 * Encoding rules:
 * - Forward slashes (/) → double underscore (__)
 * - Other filesystem-incompatible characters → _xXX_ encoding
 *   where XX is the uppercase hexadecimal representation of the character code
 *
 * Mirrors: encode_filename()
 *
 * @param {string} docId - Document identifier to encode (e.g. DOI)
 * @returns {string} Filesystem-safe encoded string
 * @throws {Error} If docId is empty
 *
 * @example
 * encodeFilename("10.1111/1467-6478.00040") // "10.1111__1467-6478.00040"
 * encodeFilename("10.1234/test:file")        // "10.1234__test_x3A_file"
 * encodeFilename("doc<name>")               // "doc_x3C_name_x3E_"
 */
export function encodeFilename(docId) {
  if (!docId) throw new Error('docId cannot be empty')

  const unsafeChars = new Set(['<', '>', ':', '"', '|', '?', '*', '\\'])

  let result = docId.replace(/\//g, '__')

  return [...result].map(char => {
    if (unsafeChars.has(char) || char.charCodeAt(0) < 32) {
      return `_x${char.charCodeAt(0).toString(16).toUpperCase().padStart(2, '0')}_`
    }
    return char
  }).join('')
}


/**
 * Decode a filesystem-safe filename back to the original document ID.
 *
 * Reverses encodeFilename():
 * - __ → / (forward slash)
 * - _xXX_ → original character (current encoding)
 * - $XX$ → original character (legacy BC encoding)
 *
 * Mirrors: decode_filename()
 *
 * @param {string} filename - Encoded filename to decode
 * @returns {string} Original document ID string
 * @throws {Error} If filename is empty
 *
 * @example
 * decodeFilename("10.1111__1467-6478.00040")  // "10.1111/1467-6478.00040"
 * decodeFilename("10.1234__test_x3A_file")    // "10.1234/test:file"
 * decodeFilename("10.1234__test$3A$file")     // "10.1234/test:file"  (legacy BC)
 */
export function decodeFilename(filename) {
  if (!filename) throw new Error('filename cannot be empty')

  let result = filename.replace(/__/g, '/')
  result = result.replace(/_x([0-9A-F]{2})_/gi, (_, hex) => String.fromCharCode(parseInt(hex, 16)))
  result = result.replace(/\$([0-9A-F]{2})\$/gi, (_, hex) => String.fromCharCode(parseInt(hex, 16)))
  return result
}


/**
 * Convert any file identifier to a valid XML NCName for use as xml:id.
 *
 * Mirrors: encode_for_xml_id()
 *
 * @param {string} fileId - Any file identifier string
 * @returns {string} NCName-safe string suitable for use as xml:id
 *
 * @example
 * encodeForXmlId("10.5771__2699-1284-2024-3-149") // "_10.5771__2699-1284-2024-3-149"
 * encodeForXmlId("test$3A$value")                 // "test_x3A_value"
 */
export function encodeForXmlId(fileId) {
  let result = normalizeLegacyEncoding(fileId)
  result = result.replace(/%([0-9A-Fa-f]{2})/g, (_, hex) => `_x${hex.toUpperCase()}_`)
  result = result.replace(/\//g, '__')
  result = result.replace(/[^a-zA-Z0-9._\-]/g, char =>
    `_x${char.charCodeAt(0).toString(16).toUpperCase().padStart(2, '0')}_`
  )
  if (result && /^\d/.test(result)) result = '_' + result
  return result
}


/**
 * Decode an xml:id value back to encode_filename() format.
 *
 * Mirrors: decode_from_xml_id()
 *
 * @param {string} xmlId - xml:id value produced by encodeForXmlId
 * @returns {string} File ID in encode_filename() format
 */
export function decodeFromXmlId(xmlId) {
  if (xmlId && xmlId.length > 1 && xmlId[0] === '_' && /\d/.test(xmlId[1])) {
    return xmlId.slice(1)
  }
  return xmlId
}


/**
 * Check if a DOI string is valid according to CrossRef specifications.
 *
 * Mirrors: validate_doi()
 *
 * @param {string} doi - The DOI string to validate
 * @returns {boolean}
 *
 * @example
 * validateDoi("10.5771/2699-1284-2024-3-149") // true
 * validateDoi("not-a-doi")                    // false
 */
export function validateDoi(doi) {
  if (!doi) return false
  return DOI_REGEX.test(doi)
}


/**
 * Normalize a DOI string: strip whitespace and common URL prefixes.
 *
 * Mirrors: normalize_doi()
 *
 * @param {string} doi - DOI string to normalize
 * @returns {string} Normalized DOI string
 *
 * @example
 * normalizeDoi("  10.5771/2699-1284-2024-3-149  ")    // "10.5771/2699-1284-2024-3-149"
 * normalizeDoi("doi:10.5771/2699-1284-2024-3-149")    // "10.5771/2699-1284-2024-3-149"
 * normalizeDoi("https://doi.org/10.5771/2699-1284-2024-3-149") // "10.5771/2699-1284-2024-3-149"
 */
export function normalizeDoi(doi) {
  if (!doi) return doi
  doi = doi.trim()
  const prefixes = [
    'doi:', 'DOI:',
    'http://doi.org/', 'https://doi.org/',
    'http://dx.doi.org/', 'https://dx.doi.org/',
  ]
  for (const prefix of prefixes) {
    if (doi.startsWith(prefix)) {
      return doi.slice(prefix.length)
    }
  }
  return doi
}


/**
 * Extract a DOI from a string. Decodes filename encoding before matching,
 * so encoded doc_ids like "10.6093__2284-0184__11598" are handled correctly.
 *
 * @param {string} str - The string to extract from
 * @returns {string|null} Extracted DOI or null
 */
export function extractDoi(str) {
  if (!str) return null
  try {
    const decoded = decodeFilename(str)
    const match = decoded.match(DOI_EXTRACT_PATTERN)
    if (match) return match[0]
  } catch (_) {
    // fall through to raw match
  }
  return str.match(DOI_EXTRACT_PATTERN)?.[0] ?? null
}


// BC alias — isDoi was the original name in utils.js
export { validateDoi as isDoi }
